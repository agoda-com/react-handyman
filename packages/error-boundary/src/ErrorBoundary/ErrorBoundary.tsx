/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */

import * as React from 'react';

const DEFAULT_ERROR = 'Error was swallowed.';

export interface ErrorFallbackProps {
  error: Error;
  errorStack: string;
}

interface ErrorBoundaryState {
  error: Error | undefined;
  errorInfo: ErrorInfo | undefined;
}

interface ErrorInfo {
  componentStack: string;
}

interface Props {
  onError: (componentName: string, error: Error, componentStack: string) => void,
  name: string,
  children?: any,
  FallbackComponent?: React.ComponentType<any>,
}

class ErrorBoundary extends React.PureComponent<Props, ErrorBoundaryState> {
  // eslint-disable-next-line  react/static-property-placement
  static displayName = 'ErrorBoundary';

  constructor(props: Props) {
    super(props);
    this.state = { error: undefined, errorInfo: undefined };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    const { onError, name } = this.props;

    if (typeof onError === 'function') {
      try {
        onError(name, error, errorInfo ? errorInfo.componentStack : 'No stack');
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error(e);
      }
    }

    this.setState({ error: error || new Error(DEFAULT_ERROR), errorInfo });
  }

  render(): JSX.Element {
    const { error, errorInfo } = this.state;
    const { children, FallbackComponent } = this.props;

    // error flow
    if (error) {
      if (FallbackComponent) {
        return <FallbackComponent error={error} errorStack={errorInfo ? errorInfo.componentStack : 'Stack not found'} />;
      }
      return <h2>Something went wrong.</h2>;
    }

    // happy flow
    return children;
  }
}

export default ErrorBoundary;
