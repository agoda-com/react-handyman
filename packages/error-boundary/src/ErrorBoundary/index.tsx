import ErrorBoundary, { ErrorFallbackProps } from './ErrorBoundary';

export default ErrorBoundary;
export type ErrorFallbackProps = ErrorFallbackProps;
