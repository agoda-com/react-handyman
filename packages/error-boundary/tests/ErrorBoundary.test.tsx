import * as React from 'react';
import { render, cleanup } from '@testing-library/react';
import ErrorBoundary, { ErrorFallbackProps } from '../src/ErrorBoundary';

describe('<ErrorBoundary />', () => {
  afterEach(() => {
    cleanup();
    jest.resetAllMocks();
  });

  it('should render fallback component on error', () => {
    class BuggyComponent extends React.PureComponent<{}, {}> {
      static displayName = 'BuggyComponent';

      constructor(props: {}) {
        super(props);
        throw Error('error');
      }

      render() {
        return <div>Dummy Component with constructor error</div>;
      }
    }

    const CustomFallbackComponent = (props: ErrorFallbackProps) => {
      const { error, errorStack } = props;

      return (
        <div>
          <h3>{error.message}</h3>
          <span>{errorStack}</span>
        </div>
      );
    };
    const errorCallback = jest.fn();

    const { asFragment } = render(
      <ErrorBoundary onError={errorCallback} name="component name" FallbackComponent={CustomFallbackComponent}>
        <BuggyComponent />
      </ErrorBoundary>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call componentDidCatch and fire callback on error', () => {
    class BuggyComponent extends React.PureComponent<{}, {}> {
      constructor(props: {}) {
        super(props);
        throw Error('error');
      }

      render() {
        return <div>Dummy Component with constructor error</div>;
      }
    }

    const errorCallback = jest.fn().mockImplementation();

    jest.spyOn(ErrorBoundary.prototype, 'componentDidCatch');

    const { asFragment } = render(
      <ErrorBoundary onError={errorCallback} name="component name">
        <BuggyComponent />
      </ErrorBoundary>
    );
    expect(ErrorBoundary.prototype.componentDidCatch).toHaveBeenCalledTimes(1);
    expect(errorCallback).toHaveBeenCalledTimes(1);
    expect(errorCallback.mock.calls[0][0]).toBe('component name');

    expect(asFragment()).toMatchSnapshot();
  });
});
