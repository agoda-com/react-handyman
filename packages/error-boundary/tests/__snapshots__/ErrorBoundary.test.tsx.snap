// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> should call componentDidCatch and fire callback on error 1`] = `
<DocumentFragment>
  <h2>
    Something went wrong.
  </h2>
</DocumentFragment>
`;

exports[`<ErrorBoundary /> should render fallback component on error 1`] = `
<DocumentFragment>
  <div>
    <h3>
      error
    </h3>
    <span>
      
    in BuggyComponent
    in ErrorBoundary
    </span>
  </div>
</DocumentFragment>
`;
